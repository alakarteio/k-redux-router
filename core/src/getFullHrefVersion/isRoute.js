export default route => !!(typeof route === 'object' && route.code)
