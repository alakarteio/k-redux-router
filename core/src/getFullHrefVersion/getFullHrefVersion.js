import { pathToRegexp, parse, compile } from 'path-to-regexp'
import isRoute from './isRoute'

export default (routes) => {
  const fullVersion = []

  const addRoute = (base, parent, route) => {
    // parameters to copy from parent to route
    let propertiesToCopy = {}
    if (parent) {
      propertiesToCopy = Object
        .entries(parent)
        // eslint-disable-next-line no-unused-vars
        .filter(([key, value]) => !isRoute(value)) // remove children routes
        .reduce(
          (acc, [key, value]) => Object.assign(acc, { [key]: value }),
          {},
        )
    }

    // process href and add to array
    let innerRoute = route
    if (parent) {
      innerRoute = Object.assign(
        propertiesToCopy,
        route,
        {
          parent: parent && parent.code,
          href: {
            base,
            compiled: (base.includes(':') ? compile(base) : undefined),
            regexp: pathToRegexp(base),
            parsed: parse(base),
          },
        },
      )

      fullVersion.push(innerRoute)
    }

    // process children
    Object
      .entries(route)
      .filter(([key, value]) => isRoute(value)) // eslint-disable-line no-unused-vars
      .forEach(([href, childRoute]) => addRoute([base, href].join('').replace('//', '/'), innerRoute, childRoute))
  }

  // start the graph
  addRoute('', undefined, routes)

  // update links
  fullVersion.forEach((route) => {
    // children
    Object
      .entries(route)
      .forEach(([key, value]) => {
        let newValue = value
        if (isRoute(value)) newValue = value.code

        route[key] = newValue // eslint-disable-line no-param-reassign
      })
  })

  return fullVersion
}
