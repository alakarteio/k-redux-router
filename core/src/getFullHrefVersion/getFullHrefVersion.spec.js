/* eslint-env jest */
import getFullHrefVersion from './getFullHrefVersion'

const match = res => () => expect(res).toMatchSnapshot()

describe('getFullHrefVersion', () => {
  it('should add no route', match(getFullHrefVersion({})))

  it('should add simples routes -nested-', match(getFullHrefVersion({
    '/': {
      code: 'main',
      '/user': {
        code: 'user',
      },
    },
  })))

  it('should add routes with path param', match(getFullHrefVersion({
    '/:id/detail': {
      code: 'user_detail',
    },
  })))

  it('should pass parameters from parent to children', match(getFullHrefVersion({
    '/': {
      code: 'main',
      public: true,
      '/user': {
        code: 'user-list',
        '/:id': {
          code: 'user',
          public: false, // override
        },
      },
    },
  })))
})
