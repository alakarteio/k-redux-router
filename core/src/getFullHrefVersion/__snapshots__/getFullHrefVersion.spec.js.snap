// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFullHrefVersion should add no route 1`] = `Array []`;

exports[`getFullHrefVersion should add routes with path param 1`] = `
Array [
  Object {
    "code": "user_detail",
    "href": Object {
      "base": "/:id/detail",
      "compiled": [Function],
      "parsed": Array [
        Object {
          "modifier": "",
          "name": "id",
          "pattern": "[^\\\\/#\\\\?]+?",
          "prefix": "/",
          "suffix": "",
        },
        "/detail",
      ],
      "regexp": /\\^\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\\\/detail\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": undefined,
  },
]
`;

exports[`getFullHrefVersion should add simples routes -nested- 1`] = `
Array [
  Object {
    "/user": "user",
    "code": "main",
    "href": Object {
      "base": "/",
      "compiled": undefined,
      "parsed": Array [
        "/",
      ],
      "regexp": /\\^\\\\/\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": undefined,
  },
  Object {
    "code": "user",
    "href": Object {
      "base": "/user",
      "compiled": undefined,
      "parsed": Array [
        "/user",
      ],
      "regexp": /\\^\\\\/user\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": "main",
  },
]
`;

exports[`getFullHrefVersion should pass parameters from parent to children 1`] = `
Array [
  Object {
    "/user": "user-list",
    "code": "main",
    "href": Object {
      "base": "/",
      "compiled": undefined,
      "parsed": Array [
        "/",
      ],
      "regexp": /\\^\\\\/\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": undefined,
    "public": true,
  },
  Object {
    "/:id": "user",
    "code": "user-list",
    "href": Object {
      "base": "/user",
      "compiled": undefined,
      "parsed": Array [
        "/user",
      ],
      "regexp": /\\^\\\\/user\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": "main",
    "public": true,
  },
  Object {
    "code": "user",
    "href": Object {
      "base": "/user/:id",
      "compiled": [Function],
      "parsed": Array [
        "/user",
        Object {
          "modifier": "",
          "name": "id",
          "pattern": "[^\\\\/#\\\\?]+?",
          "prefix": "/",
          "suffix": "",
        },
      ],
      "regexp": /\\^\\\\/user\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    },
    "parent": "user-list",
    "public": false,
  },
]
`;
